lib/record_store/provider/google_cloud_dns.rb

```diff
@@ -0,0 +1,121 @@
+        records = gcloud_record_sets.map do |record_set|
+          next if record_set.type == 'SOA'
+
+          # Unroll each record set into multiple records
+          record_set.data.map do |record|
+            begin
+              record_set_member = record_set.dup
+              record_set_member.data = [record]
+              build_from_api(record_set_member)
+            rescue StandardError => e
+              stdout.puts "Cannot build record: #{record}"
+            end
+          end
+        end
+
+        # We need to filter out for nil records (i.e. since skip the SOA record)
+        records.flatten.select(&:present?)
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        session.zones.map { |zone| zone.gapi.dns_name }
+      end
+
+      private
+
+      def session
+        @dns ||= begin
+          Google::Cloud::Dns.new({
+            project_id: secrets.fetch('project_id'),
+            credentials: Google::Cloud::Dns::Credentials.new(secrets),
+          })
+        end
+      end
+
+      def secrets
+        super.fetch('google_cloud_dns')
+      end
+
+      def convert_to_name(zone)
+        zone.gsub('.', '-')
+      end
+
+      # Google's API operates on resource record sets instead of individual
+      # records. A resource record set is a single object that has the rdata
+      # for all resource records with the same fully qualified domain name and
+      # record type. See https://cloud.google.com/dns/api/v1/resourceRecordSets
+      #
+      # This methods takes an array of records and builds Google API
+      # ResourceRecordSets objects.
+      def convert_records_to_gcloud_record_sets(zone, records)
+        record_sets = records.group_by do |record|
+          [record.type, record.fqdn]
+        end
+
+
+        record_sets.map do |(rr_type, rr_fqdn), records|
+          zone.record(rr_fqdn, rr_type, records[0].ttl, records.map(&:rdata_txt))
+        end
+      end
+
+      def build_from_api(record)
+        record_params = {
+          id: record.object_id,
+          ttl: record.ttl,
+          fqdn: record.name,
+        }
+
+        return if record.type == 'SOA'
+
+        case record.type
+        when 'A', 'AAAA'
+          record_params.merge!(address: record.data[0])
+        when 'CNAME'
+          record_params.merge!(cname: record.data[0])
+        when 'MX'
+          preference, exchange = record.data[0].split(' ')
+          record_params.merge!(preference: preference, exchange: exchange)
+        when 'NS'
+          record_params.merge!(nsdname: record.data[0])
+        when 'SPF', 'TXT'
+          record_params.merge!(txtdata: record.data[0].gsub(';', '\;'))
+        when 'SRV'
+          priority, weight, port, target = record.data[0].split(' ')
+
+          record_params.merge!(
+            priority: priority.to_i,
+            weight: weight.to_i,
+            port: port.to_i,
+            target: target,
+          )
+        end
```

Comment:
This feels like a lot of repetition, but I get why is implemented in this way. Not sure why we have different attribute names (e.g. cname, address, nsdname, txtdata) for the value of a record though.

Reply:
Yeah, it unfortunately is a bit repetitive 😞

> Not sure why we have different attribute names (e.g. cname, address, nsdname, txtdata) for the value of a record though.

This is largely based on the historical reason that the API was built around the DynECT API. I haven't re-evaluated if it makes sense to migrate to just a `rdata` string and helper accessors that extract the value. Could be a reasonable follow-up refactor.